File: _lib/python/grid.py
# -*- coding: utf-8 -*-

import sys

def input(): return sys.stdin.readline().strip()
def list2d(a, b, c): return [[c] * b for i in range(a)]
def list3d(a, b, c, d): return [[[d] * c for j in range(b)] for i in range(a)]
def list4d(a, b, c, d, e): return [[[[e] * d for j in range(c)] for j in range(b)] for i in range(a)]
def ceil(x, y=1): return int(-(-x // y))
def INT(): return int(input())
def MAP(): return map(int, input().split())
def LIST(N=None): return list(MAP()) if N is None else [INT() for i in range(N)]
def Yes(): print('Yes')
def No(): print('No')
def YES(): print('YES')
def NO(): print('NO')
sys.setrecursionlimit(10 ** 9)
INF = 10 ** 18
MOD = 10 ** 9 + 7

# 行列入れ替え
# li2 = list(zip(*li1))
# タプル->リスト変換付
# grid = [list(row) for row in list(zip(*grid))]

# 右左上下
directions = ((1, 0), (-1, 0), (0, 1), (0, -1))
# 斜め有り
directions = ((1, 0), (-1, 0), (0, 1), (0, -1), (1, -1), (-1, -1), (1, 1), (-1, 1))
# 四方に一回り大きいグリッドを作る
# grid = list2d(H+2, W+2, '*')
# for i in range(1, H+1):
#     row = list(input())
#     for j in range(1, W+1):
#         grid[i][j] = row[j-1]

def build_grid(H, W, intv, _type, space=False, padding=True):
    # 入力がスペース区切りかどうか
    if space:
        _input = lambda: input().split()
    else:
        _input = lambda: input()
    _list = lambda: list(map(_type, _input()))
    # 余白の有無
    if padding:
        offset = 1
    else:
        offset = 0
    grid = list2d(H+offset*2, W+offset*2, intv)
    for i in range(offset, H+offset):
        row = _list()
        for j in range(offset, W+offset):
            grid[i][j] = row[j-offset]
    return grid

def rot90(grid):
    """ グリッドを時計回りに90度回転 """
    H = len(grid)
    W = len(grid[0])
    res = list2d(W, H, 0)
    for i in range(H):
        for j in range(W):
            res[j][H-i-1] = grid[i][j]
    return res

def scale(grid, K):
    """ グリッドをK倍に拡大 """
    H, W = len(grid), len(grid[0])
    res = list2d(H*K, W*K, '')
    for i in range(H*K):
        for j in range(W*K):
            res[i][j] = grid[i//K][j//K]
    return res

def gridtoid(i, j, W):
    return i*W+j

def idtogrid(id, W):
    return divmod(id, W)

# ※多点スタート対応のためsrcにはlistを渡す
def bfs(grid, src):
    """ BFS(グリッド、重みなし) """
    from collections import deque

    H, W = len(grid), len(grid[0])
    directions = ((1, 0), (-1, 0), (0, 1), (0, -1))
    dist = list2d(H, W, INF)
    que = deque()
    for h, w in src:
        que.append((h, w))
        dist[h][w] = 0
    while que:
        h, w = que.popleft()
        for dh, dw in directions:
            nh = h + dh
            nw = w + dw
            if grid[nh][nw] == -1:
                continue
            if dist[nh][nw] != INF:
                continue
            dist[nh][nw] = dist[h][w] + 1
            que.append((nh, nw))
    return dist

def dijkstra(grid: list, src: tuple) -> list:
    """ グリッドダイクストラ(H*Wグリッド, 始点(h, w)) """
    from heapq import heappush, heappop

    H, W = len(grid), len(grid[0])
    # res[h][w] := マス(h, w)への最小コスト
    res = list2d(H, W, INF)
    # スタート位置
    srch, srcw = src
    que = [(0, srch, srcw)]
    res[srch][srcw] = 0
    directions = ((1, 0), (-1, 0), (0, 1), (0, -1))
    # キューが空になるまで
    while que:
        # srcからの距離, マス(h, w)
        dist, curh, curw = heappop(que)
        if res[curh][curw] < dist:
            continue
        # マスcurからの4方向でループ
        for dh, dw in directions:
            nxth, nxtw = curh + dh, curw + dw
            cost = grid[nxth][nxtw]
             # 今回の経路のが短い時
            if dist + cost < res[nxth][nxtw]:
                res[nxth][nxtw] = dist + cost
                # 現在の移動距離をキューの優先度として、早い方から先に処理するようにする
                heappush(que, (dist+cost, nxth, nxtw))
    # マスsrcからの最短距離リストを返却
    return res

File: _lib/cpp/grid.cpp
/*
・ライブラリ：グリッド
*/

#include "_tpl.cpp"


// グリッドをK倍に拡大 
vector<string> scale(const vector<string> &grid, ll K) {
    int H = grid.size();
    int W = grid[0].size();
    vector<string> res(H*K, string(W*K, '*'));
    rep(i, H*K) {
        rep(j, W*K) {
            res[i][j] = grid[i/K][j/K];
        }
    }
    return res;
}


// 8方向
const vector<pii> directions8 = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}, {-1, -1}, {-1, 1}, {1, -1}, {1, 1}};


// 斜め移動用(右下→右上→左上→左下みたいにダイヤ型に進む)
const vector<pii> directions = {{1, 1}, {-1, 1}, {-1, -1}, {1, -1}};


// グリッドBFS
vvl bfs(const vector<string> &grid, const vector<pii> &src, char invalid='#') {

    int H = grid.size();
    int W = grid[0].size();
    auto res = list2d(H, W, INF);
    const vector<pii> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    queue<pii> que;
    for (auto [h, w] : src) {
        que.push({h, w});
        res[h][w] = 0;
    }
    while (!que.empty()) {
        auto [h, w] = que.front(); que.pop();
        for (auto [dh, dw] : directions) {
            int nh = h + dh;
            int nw = w + dw;
            if (nh < 0 or nw < 0 or nh >= H or nw >= W) continue;
            if (grid[nh][nw] == invalid) continue;
            if (res[nh][nw] == INF) {
                res[nh][nw] = res[h][w] + 1;
                que.push({nh, nw});
            }
        }
    }
    return res;
}


// グリッド斜めのループ(いつもぱっと出ないでもたつくので。。)
// rep(k, -N, N) {
//     rep(i, N) {
//         if (0 <= i+k and i+k < N) {
//             // do something
//             // grid[i][i+k];
//         }
//     }
// }


// 余白付きグリッドを構築
vector<string> build_grid(int H, int W, char intv, int offset=1) {

    vector<string> res(H+offset*2);
    rep(h, offset) res[h] = string(W+offset*2, intv);
    rep(h, offset, H+offset) {
        string s;
        cin >> s;
        res[h] = string(offset, intv) + s + string(offset, intv);
    }
    rep(h, H+offset, H+offset*2) res[h] = string(W+offset*2, intv);
    return res;
}


// グリッドダイクストラ(H*Wグリッド, 始点{h, w}) 
using P = tuple<ll, int, int>;
vvl dijkstra(const vvl &grid, pii src, ll invalid=-1) {

    int H = grid.size();
    int W = grid[0].size();
    auto res = list2d(H, W, INF);
    const vector<pii> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    priority_queue<P, vector<P>, greater<P>> que;
    auto [sh, sw] = src;
    que.push({0, sh, sw});
    res[sh][sw] = 0;

    while (!que.empty()) {
        auto [dist, h, w] = que.top(); que.pop();
        if (res[h][w] < dist) continue;
        for (auto [dh, dw] : directions) {
            int nh = h+dh;
            int nw = w+dw;
            if (nh < 0 or nw < 0 or nh >= H or nw >= W) continue;
            if (grid[nh][nw] == invalid) continue;
            ll cost = grid[nh][nw];
            if (dist+cost < res[nh][nw]) {
                res[nh][nw] = dist+cost;
                que.push({dist+cost, nh, nw});
            }
        }
    }
    return res;
}
